frontend/packages/dev-console/src/components/topology/topology-utils.ts

```diff
@@ -24,412 +30,247 @@ export const getEditURL = (gitURL: string, cheURL: string) => {
-          isKnativeResource: this.transformPodData.isKnativeServing(
-            deploymentConfig,
-            'metadata.labels',
-          ),
-          donutStatus: {
-            pods: [...currentPods, ...previousPods],
-            build: builds && builds[0],
-          },
-        },
-      };
-    });
-    return this;
-  }
+/**
+ * get the route data
+ */
+const getRouteData = (ksroute: K8sResourceKind[]): string => {
+  return ksroute && !_.isEmpty(ksroute[0].status) ? ksroute[0].status.url : null;
+};
 
-  private getRoutesUrl(routes: ResourceProps[], ksroute: ResourceProps): string {
-    if (routes.length > 0 && !_.isEmpty(routes[0].spec)) {
-      return getRouteWebURL(routes[0]);
-    }
-    return this.getRouteData(ksroute);
+/**
+ * get routes url
+ */
+const getRoutesUrl = (routes: K8sResourceKind[], ksroute?: K8sResourceKind[]): string => {
+  if (routes.length > 0 && !_.isEmpty(routes[0].spec)) {
+    return getRouteWebURL(routes[0]);
   }
+  return getRouteData(ksroute);
+};
 
-  private getSelectorsByService() {
-    let allServices = _.keyBy(this.resources.services.data, 'metadata.name');
-    if (this.resources.ksservices && this.resources.ksservices.data) {
-      allServices = _.keyBy(
-        [...this.resources.services.data, ...this.resources.ksservices.data],
-        'metadata.name',
-      );
-    }
-    const selectorsByService = _.mapValues(allServices, (service) => {
-      return new LabelSelector(service.spec.selector);
-    });
-    return selectorsByService;
-  }
+/**
+ * create instance of TransformResourceData, return object containing all methods
+ * @param resources
+ * @param utils
+ */
+const createInstanceForResource = (resources: TopologyDataResources, utils?: Function[]) => {
+  const transformResourceData = new TransformResourceData(resources, utils);
+
+  return {
+    deployments: transformResourceData.createDeploymentItems,
+    deploymentConfigs: transformResourceData.createDeploymentConfigItems,
+    daemonSets: transformResourceData.createDaemonSetItems,
+    statefulSets: transformResourceData.createStatefulSetItems,
+  };
+};
 
-  /**
-   * filter data based on the active application
-   * @param data
-   */
-  private filterBasedOnActiveApplication(data) {
-    const PART_OF = 'app.kubernetes.io/part-of';
-    if (!this.application) {
-      return data;
-    }
-    return data.filter((dc) => {
-      return _.get(dc, ['metadata', 'labels', PART_OF]) === this.application;
-    });
-  }
+/**
+ * create all data that need to be shown on a topology data
+ * @param dc resource item
+ * @param cheURL che link
+ */
+const createTopologyNodeData = (dc: OverviewItem, cheURL?: string): TopologyDataObject => {
+  const { obj: deploymentConfig } = dc;
+  const dcUID = _.get(deploymentConfig, 'metadata.uid');
+  const deploymentsLabels = _.get(deploymentConfig, 'metadata.labels') || {};
+  const deploymentsAnnotations = _.get(deploymentConfig, 'metadata.annotations') || {};
```

Comment:
`_.get` possesses a 3rd param for defaulting the value when not found (aka `undefined`). Is there concern that `labels` or `annotations` are going to be there but not objects (ie `null`)?

https://lodash.com/docs/4.17.15#get

Reply:
fixed in https://github.com/openshift/console/pull/2849/commits/116a1f08ce828a61035940e67d4991ea8918da7b